<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>opqua.internal package &mdash; Opqua 1.0.2 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=1ed6394b"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="opqua package" href="opqua.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Opqua
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="requirements_and_installation.html">Requirements and Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_documentation.html"><code class="docutils literal notranslate"><span class="pre">Model</span></code> Documentation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="API.html">API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="opqua.html">opqua package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="opqua.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4 current"><a class="current reference internal" href="#">opqua.internal package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="opqua.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="opqua.html#module-opqua.model">opqua.model module</a></li>
<li class="toctree-l3"><a class="reference internal" href="opqua.html#module-opqua">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Opqua</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="API.html">API</a></li>
          <li class="breadcrumb-item"><a href="opqua.html">opqua package</a></li>
      <li class="breadcrumb-item active">opqua.internal package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/opqua.internal.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="opqua-internal-package">
<h1>opqua.internal package<a class="headerlink" href="#opqua-internal-package" title="Link to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading"></a></h2>
</section>
<section id="module-opqua.internal.data">
<span id="opqua-internal-data-module"></span><h2>opqua.internal.data module<a class="headerlink" href="#module-opqua.internal.data" title="Link to this heading"></a></h2>
<p>Contains data wrangling methods.</p>
<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.data.compartmentDf">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.data.</span></span><span class="sig-name descname"><span class="pre">compartmentDf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hosts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vectors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_to_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/data.html#compartmentDf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.data.compartmentDf" title="Link to this definition"></a></dt>
<dd><p>Create dataframe with number of naive, susc., inf., rec. hosts/vectors.</p>
<p>Creates a pandas Dataframe with dynamics of all compartments (naive,
infected, recovered, dead) across selected populations in the model,
with one time point in each row and columns for time as well as each
compartment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pandas DataFrame</em>) – dataframe with model history as produced by <cite>saveToDf</cite> function.</p>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>populations</strong> (<em>list</em><em> of </em><em>Strings</em>) – IDs of populations to include in analysis; if empty, uses all
populations in model. Defaults to [].</p></li>
<li><p><strong>hosts</strong> (<em>Boolean</em>) – whether to count hosts. Defaults to True.</p></li>
<li><p><strong>vectors</strong> (<em>Boolean</em>) – whether to count vectors. Defaults to False.</p></li>
<li><p><strong>save_to_file</strong> (<em>String</em>) – file path and name to save model data under, no saving
occurs if empty string. Defaults to “”.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas DataFrame with model compartment dynamics as described above.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.data.compositionDf">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.data.</span></span><span class="sig-name descname"><span class="pre">compositionDf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_of_composition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Pathogens'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hosts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vectors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_top_sequences</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">track_specific_sequences</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genomic_positions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count_individuals_based_on_model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_to_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/data.html#compositionDf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.data.compositionDf" title="Link to this definition"></a></dt>
<dd><p>Create dataframe with counts for pathogen genomes or resistance.</p>
<p>Creates a pandas Dataframe with dynamics of the pathogen strains or
protection sequences across selected populations in the model,
with one time point in each row and columns for pathogen genomes or
protection sequences.</p>
<p>Of note: sum of totals for all sequences in one time point does not
necessarily equal the number of infected hosts and/or vectors, given
multiple infections in the same host/vector are counted separately.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pandas DataFrame</em>) – dataframe with model history as produced by <cite>saveToDf</cite> function.</p>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>populations</strong> (<em>list</em><em> of </em><em>Strings</em>) – IDs of populations to include in analysis; if empty, uses all
populations in model. Defaults to [].</p></li>
<li><p><strong>type_of_composition</strong> (<em>String</em>) – field of data to count totals of, can be either
‘Pathogens’ or ‘Protection’. Defaults to ‘Pathogens’.</p></li>
<li><p><strong>hosts</strong> (<em>Boolean</em>) – whether to count hosts. Defaults to True.</p></li>
<li><p><strong>vectors</strong> (<em>Boolean</em>) – whether to count vectors. Defaults to False.</p></li>
<li><p><strong>num_top_sequences</strong> (<em>int</em>) – how many sequences to count separately and include
as columns, remainder will be counted under column “Other”; if &lt;0,
includes all genomes in model. Defaults to -1.</p></li>
<li><p><strong>track_specific_sequences</strong> (<em>list</em><em> of </em><em>Strings</em>) – contains specific sequences to have
as a separate column if not part of the top num_top_sequences
sequences. Defaults to [].</p></li>
<li><p><strong>genomic_positions</strong> (<em>list</em><em> of </em><em>lists</em><em> of </em><em>int</em>) – list in which each element is a list with loci
positions to extract (e.g. genomic_positions=[ [0,3], [5,6] ] extracts
positions 0, 1, 2, and 5 from each genome); if empty, takes full genomes. Defaults to [].</p></li>
<li><p><strong>count_individuals_based_on_model</strong> (<em>None</em><em> or </em><a class="reference internal" href="opqua.html#opqua.model.Model" title="opqua.model.Model"><em>Model</em></a>) – Model object with populations and
fitness functions used to evaluate the most fit pathogen genome in each
host/vector in order to count only a single pathogen per host/vector, as
opposed to all pathogens within each host/vector; if None, counts all
pathogens. Defaults to None.</p></li>
<li><p><strong>save_to_file</strong> (<em>String</em>) – file path and name to save model data under, no saving
occurs if empty string. Defaults to “”.</p></li>
<li><p><strong>n_cores</strong> (<em>int</em>) – number of cores to parallelize processing across, if 0, all
cores available are used. Defaults to 0.</p></li>
<li><p><strong>**kwargs</strong> – additional arguents for joblib multiprocessing.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas DataFrame with model sequence composition dynamics as described above.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.data.getGenomeTimesDf">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.data.</span></span><span class="sig-name descname"><span class="pre">getGenomeTimesDf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_to_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/data.html#getGenomeTimesDf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.data.getGenomeTimesDf" title="Link to this definition"></a></dt>
<dd><p>Create DataFrame with times genomes first appeared during simulation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pandas DataFrame</em>) – dataframe with model history as produced by <cite>saveToDf</cite> function.</p>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>samples</strong> (<em>int</em>) – how many timepoints to uniformly sample from the total
timecourse; if &lt;0, takes all timepoints. Defaults to 1.</p></li>
<li><p><strong>save_to_file</strong> (<em>String</em>) – file path and name to save model data under, no saving
occurs if empty string. Defaults to “”.</p></li>
<li><p><strong>n_cores</strong> (<em>int</em>) – number of cores to parallelize across, if 0, all cores available
are used. Defaults to 0.</p></li>
<li><p><strong>**kwargs</strong> – additional arguents for joblib multiprocessing.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas DataFrame with genomes and times as described above.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.data.getPathogenDistanceHistoryDf">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.data.</span></span><span class="sig-name descname"><span class="pre">getPathogenDistanceHistoryDf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_top_sequences</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">track_specific_sequences</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_to_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/data.html#getPathogenDistanceHistoryDf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.data.getPathogenDistanceHistoryDf" title="Link to this definition"></a></dt>
<dd><p>Create DataFrame with pairwise Hamming distances for pathogen sequences
in data.</p>
<p>DataFrame has indexes and columns named according to genomes or argument
<cite>seq_names</cite>, if passed. Distance is measured as percent Hamming distance from
an optimal genome sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pandas DataFrame</em>) – dataframe with model history as produced by <cite>saveToDf</cite> function</p>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>samples</strong> (<em>int</em>) – how many timepoints to uniformly sample from the total
timecourse; if &lt;0, takes all timepoints. Defaults to 1.</p></li>
<li><p><strong>num_top_sequences</strong> (<em>int</em>) – how many sequences to include in matrix; if &lt;0,
includes all genomes in data passed. Defaults to -1.</p></li>
<li><p><strong>track_specific_sequences</strong> (<em>list</em><em> of </em><em>Strings</em>) – contains specific sequences to include in matrix
if not part of the top num_top_sequences sequences. Defaults to [].</p></li>
<li><p><strong>seq_names</strong> (<em>list</em><em> of </em><em>Strings</em>) – list with names to be used for sequence labels in matrix must
be of same length as number of sequences to be displayed; if empty,
uses sequences themselves. Defaults to [].</p></li>
<li><p><strong>save_to_file</strong> (<em>String</em>) – file path and name to save model data under, no saving
occurs if empty string. Defaults to “”.</p></li>
<li><p><strong>n_cores</strong> (<em>int</em>) – number of cores to parallelize distance compute across, if 0, all
cores available are used (default 0; int)</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas DataFrame with distance matrix as described above.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.data.getPathogens">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.data.</span></span><span class="sig-name descname"><span class="pre">getPathogens</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_to_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/data.html#getPathogens"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.data.getPathogens" title="Link to this definition"></a></dt>
<dd><p>Create Dataframe with counts for all pathogen genomes in data.</p>
<p>Returns sorted pandas DataFrame with counts for occurrences of all pathogen
genomes in data passed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pandas DataFrame</em>) – dataframe with model history as produced by <cite>saveToDf</cite> function.</p>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>save_to_file</strong> (<em>String</em>) – file path and name to save model data under, no saving
occurs if empty string. Defaults to “”.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas DataFrame with Series as described above.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.data.getProtections">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.data.</span></span><span class="sig-name descname"><span class="pre">getProtections</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_to_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/data.html#getProtections"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.data.getProtections" title="Link to this definition"></a></dt>
<dd><p>Create Dataframe with counts for all protection sequences in data.</p>
<p>Returns sorted pandas DataFrame with counts for occurrences of all
protection sequences in data passed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pandas DataFrame</em>) – dataframe with model history as produced by <cite>saveToDf</cite> function.</p>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>save_to_file</strong> (<em>String</em>) – file path and name to save model data under, no saving
occurs if empty string. Defaults to “”.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas DataFrame with Series as described above.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.data.pathogenDistanceDf">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.data.</span></span><span class="sig-name descname"><span class="pre">pathogenDistanceDf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_top_sequences</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">track_specific_sequences</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_to_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/data.html#pathogenDistanceDf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.data.pathogenDistanceDf" title="Link to this definition"></a></dt>
<dd><p>Create DataFrame with pairwise Hamming distances for pathogen sequences
in data.</p>
<p>DataFrame has indexes and columns named according to genomes or argument
<cite>seq_names</cite>, if passed. Distance is measured as percent Hamming distance from
an optimal genome sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pandas DataFrame</em>) – dataframe with model history as produced by <cite>saveToDf</cite> function.</p>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>num_top_sequences</strong> (<em>int</em>) – how many sequences to include in matrix; if &lt;0,
includes all genomes in data passed. Defaults to -1.</p></li>
<li><p><strong>track_specific_sequences</strong> (<em>list</em><em> of </em><em>Strings</em>) – contains specific sequences to include in matrix
if not part of the top num_top_sequences sequences. Defaults to [].</p></li>
<li><p><strong>seq_names</strong> (<em>list</em><em> of </em><em>Strings</em>) – list with names to be used for sequence labels in matrix must
be of same length as number of sequences to be displayed; if empty,
uses sequences themselves. Defaults to [].</p></li>
<li><p><strong>save_to_file</strong> (<em>String</em>) – file path and name to save model data under, no saving
occurs if empty string. Defaults to “”.</p></li>
<li><p><strong>n_cores</strong> (<em>int</em>) – number of cores to parallelize distance compute across, if 0, all
cores available are used. Defaults to 0.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas DataFrame with distance matrix as described above.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.data.populationsDf">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.data.</span></span><span class="sig-name descname"><span class="pre">populationsDf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compartment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Infected'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hosts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vectors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_top_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">track_specific_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_to_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/data.html#populationsDf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.data.populationsDf" title="Link to this definition"></a></dt>
<dd><p>Create dataframe with aggregated totals per population.</p>
<p>Creates a pandas Dataframe in long format with dynamics of a compartment
across populations in the model, with one time point in each row and columns
for time as well as each population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pandas DataFrame</em>) – </p>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>compartment</strong> (<em>String</em>) – subset of hosts/vectors to count totals of, can be either
‘Naive’,’Infected’,’Recovered’, or ‘Dead’. Defaults to ‘Infected’.</p></li>
<li><p><strong>hosts</strong> (<em>Boolean</em>) – whether to count hosts. Defaults to True.</p></li>
<li><p><strong>vectors</strong> (<em>Boolean</em>) – whether to count vectors. Defaults to False.</p></li>
<li><p><strong>num_top_populations</strong> (<em>int</em>) – how many populations to count separately and include
as columns, remainder will be counted under column “Other”; if &lt;0,
includes all populations in model. Defaults to -1.</p></li>
<li><p><strong>track_specific_populations</strong> (<em>list</em><em> of </em><em>Strings</em>) – contains IDs of specific populations to have
as a separate column if not part of the top num_top_populations
populations. Defaults to [].</p></li>
<li><p><strong>save_to_file</strong> (<em>String</em>) – file path and name to save model data under, no saving
occurs if empty string. Defaults to “”.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas DataFrame with model population dynamics as described above.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.data.saveToDf">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.data.</span></span><span class="sig-name descname"><span class="pre">saveToDf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">history</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_to_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/data.html#saveToDf"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.data.saveToDf" title="Link to this definition"></a></dt>
<dd><p>Save status of model to dataframe, write to file location given.</p>
<p>Creates a pandas Dataframe in long format with the given model history, with
one host or vector per simulation time in each row, and columns:</p>
<blockquote>
<div><ul class="simple">
<li><p>Time - simulation time of entry</p></li>
<li><p>Population - ID of this host/vector’s population</p></li>
<li><p>Organism - host/vector</p></li>
<li><p>ID - ID of host/vector</p></li>
<li><p>Pathogens - all genomes present in this host/vector separated by ‘;’</p></li>
<li><p>Protection - all genomes present in this host/vector separated by ‘;’</p></li>
<li><p>Alive - whether host/vector is alive at this time, True/False</p></li>
</ul>
</div></blockquote>
<p>Writing straight to a file and then reading into a pandas dataframe was
actually more efficient than concatenating directly into a pd dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>history</strong> (<em>dict</em>) – dictionary containing model state history, with <cite>keys`=`times</cite> and
<cite>values`=`Model</cite> objects with model snapshot at that time point.</p></li>
<li><p><strong>save_to_file</strong> (<em>String</em>) – file path and name to save model data under.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>n_cores</strong> (<em>int</em>) – number of cores to parallelize file export across, if 0, all
cores available are used. Defaults to 0.</p></li>
<li><p><strong>**kwargs</strong> – additional arguents for joblib multiprocessing.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas DataFrame with model history as described above.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-opqua.internal.gillespie">
<span id="opqua-internal-gillespie-module"></span><h2>opqua.internal.gillespie module<a class="headerlink" href="#module-opqua.internal.gillespie" title="Link to this heading"></a></h2>
<p>Contains class Population.</p>
<dl class="py class">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">opqua.internal.gillespie.</span></span><span class="sig-name descname"><span class="pre">Gillespie</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/gillespie.html#Gillespie"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.gillespie.Gillespie" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class contains methods for simulating model with Gillespie algorithm.</p>
<p>Class defines a model’s events and methods for changing system state
according to the possible events and simulating a timecourse using the
Gillespie algorithm.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.model">
<span class="sig-name descname"><span class="pre">model</span></span><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.model" title="Link to this definition"></a></dt>
<dd><p>the model this simulation belongs to.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Model object</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.BIRTH_HOST">
<span class="sig-name descname"><span class="pre">BIRTH_HOST</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">18</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.BIRTH_HOST" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.BIRTH_VECTOR">
<span class="sig-name descname"><span class="pre">BIRTH_VECTOR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">19</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.BIRTH_VECTOR" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.CONTACT_HOST_HOST">
<span class="sig-name descname"><span class="pre">CONTACT_HOST_HOST</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">5</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.CONTACT_HOST_HOST" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.CONTACT_HOST_VECTOR">
<span class="sig-name descname"><span class="pre">CONTACT_HOST_VECTOR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">6</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.CONTACT_HOST_VECTOR" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.CONTACT_VECTOR_HOST">
<span class="sig-name descname"><span class="pre">CONTACT_VECTOR_HOST</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">7</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.CONTACT_VECTOR_HOST" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.DIE_HOST">
<span class="sig-name descname"><span class="pre">DIE_HOST</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">16</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.DIE_HOST" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.DIE_VECTOR">
<span class="sig-name descname"><span class="pre">DIE_VECTOR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">17</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.DIE_VECTOR" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.EVENT_IDS">
<span class="sig-name descname"><span class="pre">EVENT_IDS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{0:</span> <span class="pre">'MIGRATE_HOST',</span> <span class="pre">1:</span> <span class="pre">'MIGRATE_VECTOR',</span> <span class="pre">2:</span> <span class="pre">'POPULATION_CONTACT_HOST_HOST',</span> <span class="pre">3:</span> <span class="pre">'POPULATION_CONTACT_HOST_VECTOR',</span> <span class="pre">4:</span> <span class="pre">'POPULATION_CONTACT_VECTOR_HOST',</span> <span class="pre">5:</span> <span class="pre">'CONTACT_HOST_HOST',</span> <span class="pre">6:</span> <span class="pre">'CONTACT_HOST_VECTOR',</span> <span class="pre">7:</span> <span class="pre">'CONTACT_VECTOR_HOST',</span> <span class="pre">8:</span> <span class="pre">'RECOVER_HOST',</span> <span class="pre">9:</span> <span class="pre">'RECOVER_VECTOR',</span> <span class="pre">10:</span> <span class="pre">'MUTATE_HOST',</span> <span class="pre">11:</span> <span class="pre">'MUTATE_VECTOR',</span> <span class="pre">12:</span> <span class="pre">'RECOMBINE_HOST',</span> <span class="pre">13:</span> <span class="pre">'RECOMBINE_VECTOR',</span> <span class="pre">14:</span> <span class="pre">'KILL_HOST',</span> <span class="pre">15:</span> <span class="pre">'KILL_VECTOR',</span> <span class="pre">16:</span> <span class="pre">'DIE_HOST',</span> <span class="pre">17:</span> <span class="pre">'DIE_VECTOR',</span> <span class="pre">18:</span> <span class="pre">'BIRTH_HOST',</span> <span class="pre">19:</span> <span class="pre">'BIRTH_VECTOR'}</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.EVENT_IDS" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.KILL_HOST">
<span class="sig-name descname"><span class="pre">KILL_HOST</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">14</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.KILL_HOST" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.KILL_VECTOR">
<span class="sig-name descname"><span class="pre">KILL_VECTOR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">15</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.KILL_VECTOR" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.MIGRATE_HOST">
<span class="sig-name descname"><span class="pre">MIGRATE_HOST</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">0</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.MIGRATE_HOST" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.MIGRATE_VECTOR">
<span class="sig-name descname"><span class="pre">MIGRATE_VECTOR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.MIGRATE_VECTOR" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.MUTATE_HOST">
<span class="sig-name descname"><span class="pre">MUTATE_HOST</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">10</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.MUTATE_HOST" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.MUTATE_VECTOR">
<span class="sig-name descname"><span class="pre">MUTATE_VECTOR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">11</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.MUTATE_VECTOR" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.POPULATION_CONTACT_HOST_HOST">
<span class="sig-name descname"><span class="pre">POPULATION_CONTACT_HOST_HOST</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">2</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.POPULATION_CONTACT_HOST_HOST" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.POPULATION_CONTACT_HOST_VECTOR">
<span class="sig-name descname"><span class="pre">POPULATION_CONTACT_HOST_VECTOR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">3</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.POPULATION_CONTACT_HOST_VECTOR" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.POPULATION_CONTACT_VECTOR_HOST">
<span class="sig-name descname"><span class="pre">POPULATION_CONTACT_VECTOR_HOST</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">4</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.POPULATION_CONTACT_VECTOR_HOST" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.RECOMBINE_HOST">
<span class="sig-name descname"><span class="pre">RECOMBINE_HOST</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">12</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.RECOMBINE_HOST" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.RECOMBINE_VECTOR">
<span class="sig-name descname"><span class="pre">RECOMBINE_VECTOR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">13</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.RECOMBINE_VECTOR" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.RECOVER_HOST">
<span class="sig-name descname"><span class="pre">RECOVER_HOST</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">8</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.RECOVER_HOST" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.RECOVER_VECTOR">
<span class="sig-name descname"><span class="pre">RECOVER_VECTOR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">9</span></em><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.RECOVER_VECTOR" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.doAction">
<span class="sig-name descname"><span class="pre">doAction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">act</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/gillespie.html#Gillespie.doAction"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.doAction" title="Link to this definition"></a></dt>
<dd><p>Change system state according to act argument passed</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>act</strong> (<em>int</em>) – defines action to be taken, one of the event ID constants.</p></li>
<li><p><strong>pop</strong> (<em>Population object</em>) – where the population action will happen in.</p></li>
<li><p><strong>rand</strong> (<em>number 0-1</em>) – random number used to define event.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Boolean indicationg whether or not the model has changed state.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.getRates">
<span class="sig-name descname"><span class="pre">getRates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">population_ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/gillespie.html#Gillespie.getRates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.getRates" title="Link to this definition"></a></dt>
<dd><p>Wrapper for calculating event rates as per current system state.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>population_ids</strong> (<em>list</em><em> of </em><em>Strings</em>) – list with IDs for every population in the model.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Matrix with rates as values for events (rows) and populations (columns).
Populations in order given in argument.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.gillespie.Gillespie.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tf</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_sampling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">host_sampling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vector_sampling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">print_every_n_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/gillespie.html#Gillespie.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.gillespie.Gillespie.run" title="Link to this definition"></a></dt>
<dd><p>Simulate model for a specified time between two time points.</p>
<p>Simulates a time series using the Gillespie algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>t0</strong> (<em>number</em>) – initial time point to start simulation at.</p></li>
<li><p><strong>tf</strong> (<em>number</em>) – initial time point to end simulation at.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>time_sampling</strong> (<em>int</em>) – how many events to skip before saving a snapshot of the
system state (saves all by default), if &lt;0, saves only final state. Defaults to 0.</p></li>
<li><p><strong>host_sampling</strong> (<em>int</em>) – how many hosts to skip before saving one in a snapshot
of the system state (saves all by default). Defaults to 0.</p></li>
<li><p><strong>vector_sampling</strong> (<em>int</em>) – how many vectors to skip before saving one in a
snapshot of the system state (saves all by default). Defaults to 0.</p></li>
<li><p><strong>print_every_n_events</strong> (<em>int&gt;0</em>) – number of events a message is printed to console. Defaults to 1000.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>dictionary containing model state history, with <cite>keys`=`times</cite> and</dt><dd><p><cite>values`=`Model</cite> objects with model snapshot at that time point.</p>
</dd>
</dl>
</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-opqua.internal.host">
<span id="opqua-internal-host-module"></span><h2>opqua.internal.host module<a class="headerlink" href="#module-opqua.internal.host" title="Link to this heading"></a></h2>
<p>Contains class Host.</p>
<dl class="py class">
<dt class="sig sig-object py" id="opqua.internal.host.Host">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">opqua.internal.host.</span></span><span class="sig-name descname"><span class="pre">Host</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">population</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class defines main entities to be infected by pathogens in model.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.host.Host.population">
<span class="sig-name descname"><span class="pre">population</span></span><a class="headerlink" href="#opqua.internal.host.Host.population" title="Link to this definition"></a></dt>
<dd><p>the population this host belongs to.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Population object</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.host.Host.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#opqua.internal.host.Host.id" title="Link to this definition"></a></dt>
<dd><p>unique identifier for this host within population.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>String</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.host.Host.slim">
<span class="sig-name descname"><span class="pre">slim</span></span><a class="headerlink" href="#opqua.internal.host.Host.slim" title="Link to this definition"></a></dt>
<dd><p>whether to create a slimmed-down representation of the
population for data storage (only ID, host and vector lists). Defaults to
False.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Boolean</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.host.Host.acquirePathogen">
<span class="sig-name descname"><span class="pre">acquirePathogen</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host.acquirePathogen"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host.acquirePathogen" title="Link to this definition"></a></dt>
<dd><p>Adds given genome to this host’s pathogens.</p>
<p>Modifies event coefficient matrix accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>genome</strong> (<em>String</em>) – the genome to be added.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.host.Host.applyTreatment">
<span class="sig-name descname"><span class="pre">applyTreatment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resistance_seqs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host.applyTreatment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host.applyTreatment" title="Link to this definition"></a></dt>
<dd><p>Remove all infections with genotypes susceptible to given treatment.</p>
<p>Pathogens are removed if they are missing at least one of the sequences
in resistance_seqs from their genome. Removes this organism from
population infected list and adds to healthy list if appropriate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>resistance_seqs</strong> (<em>list</em><em> of </em><em>Strings</em>) – contains sequences required for treatment resistance.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.host.Host.birth">
<span class="sig-name descname"><span class="pre">birth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host.birth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host.birth" title="Link to this definition"></a></dt>
<dd><p>Add a new host to population based on this host.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.host.Host.copyState">
<span class="sig-name descname"><span class="pre">copyState</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host.copyState"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host.copyState" title="Link to this definition"></a></dt>
<dd><p>Returns a slimmed-down representation of the current host state.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Host object with current pathogens and protection_sequences.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.host.Host.die">
<span class="sig-name descname"><span class="pre">die</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host.die"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host.die" title="Link to this definition"></a></dt>
<dd><p>Add host to population’s dead list, remove it from alive ones.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.host.Host.getWeightedRandomGenome">
<span class="sig-name descname"><span class="pre">getWeightedRandomGenome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host.getWeightedRandomGenome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host.getWeightedRandomGenome" title="Link to this definition"></a></dt>
<dd><p>Returns index of element chosen from weights and given random number.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rand</strong> (<em>number 0-1</em>) – random number.</p></li>
<li><p><strong>r</strong> (<em>numpy array</em>) – array with weights.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>new 0-1 random number.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.host.Host.infectHost">
<span class="sig-name descname"><span class="pre">infectHost</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">host</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host.infectHost"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host.infectHost" title="Link to this definition"></a></dt>
<dd><p>Infect given host with a sample of this host’s pathogens.</p>
<p>Each pathogen in the infector is sampled as present or absent in the
inoculum by drawing from a Poisson distribution with a mean equal to the
mean inoculum size of the organism being infected weighted by each
genome’s fitness as a fraction of the total in the infector as the
probability of each trial (minimum 1 pathogen transfered). Each pathogen
present in the inoculum will be added to the infected organism, if it
does not have protection from the pathogen’s genome. Fitnesses are
computed for the pathogens’ genomes in the infected organism, and the
organism is included in the poplation’s infected list if appropriate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vector</strong> (<em>Vector object</em>) – the vector to be infected.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Boolean indicating whether or not the model has changed state.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.host.Host.infectVector">
<span class="sig-name descname"><span class="pre">infectVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host.infectVector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host.infectVector" title="Link to this definition"></a></dt>
<dd><p>Infect given host with a sample of this host’s pathogens.</p>
<p>Each pathogen in the infector is sampled as present or absent in the
inoculum by drawing from a Poisson distribution with a mean equal to the
mean inoculum size of the organism being infected weighted by each
genome’s fitness as a fraction of the total in the infector as the
probability of each trial (minimum 1 pathogen transfered). Each pathogen
present in the inoculum will be added to the infected organism, if it
does not have protection from the pathogen’s genome. Fitnesses are
computed for the pathogens’ genomes in the infected organism, and the
organism is included in the poplation’s infected list if appropriate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vector</strong> (<em>Vector object</em>) – the vector to be infected.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Boolean indicating whether or not the model has changed state.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.host.Host.mutate">
<span class="sig-name descname"><span class="pre">mutate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host.mutate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host.mutate" title="Link to this definition"></a></dt>
<dd><p>Mutate a single, random locus in a random pathogen.</p>
<p>Creates a new genotype from a de novo mutation event.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.host.Host.recombine">
<span class="sig-name descname"><span class="pre">recombine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host.recombine"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host.recombine" title="Link to this definition"></a></dt>
<dd><p>Recombine two random pathogen genomes at random locus.</p>
<p>Creates a new genotype from two random possible pathogens.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.host.Host.recover">
<span class="sig-name descname"><span class="pre">recover</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/host.html#Host.recover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.host.Host.recover" title="Link to this definition"></a></dt>
<dd><p>Remove all infections from this host.</p>
<p>If model is protecting upon recovery, add protecion sequence as defined
by the indexes in the corresponding model parameter. Remove from
population infected list and add to healthy list.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-opqua.internal.intervention">
<span id="opqua-internal-intervention-module"></span><h2>opqua.internal.intervention module<a class="headerlink" href="#module-opqua.internal.intervention" title="Link to this heading"></a></h2>
<p>Contains class Intervention.</p>
<dl class="py class">
<dt class="sig sig-object py" id="opqua.internal.intervention.Intervention">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">opqua.internal.intervention.</span></span><span class="sig-name descname"><span class="pre">Intervention</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">time</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/intervention.html#Intervention"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.intervention.Intervention" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class defines a new intervention to be done at a specified time.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.intervention.Intervention.time">
<span class="sig-name descname"><span class="pre">time</span></span><a class="headerlink" href="#opqua.internal.intervention.Intervention.time" title="Link to this definition"></a></dt>
<dd><p>time at which intervention will take place.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.intervention.Intervention.method_name">
<span class="sig-name descname"><span class="pre">method_name</span></span><a class="headerlink" href="#opqua.internal.intervention.Intervention.method_name" title="Link to this definition"></a></dt>
<dd><p>intervention to be carried out, must correspond to the
name of a method of the Model object.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>String</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.intervention.Intervention.args">
<span class="sig-name descname"><span class="pre">args</span></span><a class="headerlink" href="#opqua.internal.intervention.Intervention.args" title="Link to this definition"></a></dt>
<dd><p>contains arguments for function in positinal order.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>array-like</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.intervention.Intervention.model">
<span class="sig-name descname"><span class="pre">model</span></span><a class="headerlink" href="#opqua.internal.intervention.Intervention.model" title="Link to this definition"></a></dt>
<dd><p>Model object this intervention is associated to.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Model object</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.intervention.Intervention.doIntervention">
<span class="sig-name descname"><span class="pre">doIntervention</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/intervention.html#Intervention.doIntervention"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.intervention.Intervention.doIntervention" title="Link to this definition"></a></dt>
<dd><p>Execute intervention function with specified arguments.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-opqua.internal.plot">
<span id="opqua-internal-plot-module"></span><h2>opqua.internal.plot module<a class="headerlink" href="#module-opqua.internal.plot" title="Link to this heading"></a></h2>
<p>Contains graphmaking methods.</p>
<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.plot.clustermap">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.plot.</span></span><span class="sig-name descname"><span class="pre">clustermap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_top_sequences=-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">track_specific_sequences=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_names=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cores=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method='weighted'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric='euclidean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_data_to_file=''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_title='Distance'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_values=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize=(10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">10)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi=200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_map=&lt;matplotlib.colors.ListedColormap</span> <span class="pre">object&gt;</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/plot.html#clustermap"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.plot.clustermap" title="Link to this definition"></a></dt>
<dd><p>Create a heatmap and dendrogram for pathogen genomes in data passed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> (<em>String</em>) – file path, name, and extension to save plot under.</p></li>
<li><p><strong>data</strong> (<em>pandas DataFrame</em>) – dataframe with model history as produced by <cite>saveToDf</cite> function.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>num_top_sequences</strong> (<em>int</em>) – how many sequences to include in matrix; if &lt;0,
includes all genomes in data passed. Deafults to -1.</p></li>
<li><p><strong>track_specific_sequences</strong> (<em>list</em><em> of </em><em>Strings</em>) – contains specific sequences to include in matrix
if not part of the top num_top_sequences sequences. Defaults to [].</p></li>
<li><p><strong>seq_names</strong> (<em>list</em><em> of </em><em>Strings</em>) – list with names to be used for sequence labels in matrix must
be of same length as number of sequences to be displayed; if empty,
uses sequences themselves. Defaults to [].</p></li>
<li><p><strong>n_cores</strong> (<em>int</em>) – number of cores to parallelize distance compute across, if 0, all
cores available are used. Defaults to 0.</p></li>
<li><p><strong>method</strong> (<em>String</em>) – clustering algorithm to use with seaborn clustermap. Defaults to ‘weighted’.</p></li>
<li><p><strong>metric</strong> (<em>String</em>) – distance metric to use with seaborn clustermap. Defaults to ‘euclidean’.</p></li>
<li><p><strong>save_data_to_file</strong> (<em>String</em>) – file path and name to save model data under, no saving
occurs if empty string. Defaults to “”.</p></li>
<li><p><strong>legend_title</strong> (<em>String</em>) – legend title. Defaults to ‘Distance’.</p></li>
<li><p><strong>figsize</strong> (<em>array-like</em><em> of </em><em>two ints</em>) – dimensions of figure. Defaults to (8,4).</p></li>
<li><p><strong>dpi</strong> (<em>int</em>) – figure resolution. Defaults to 200.</p></li>
<li><p><strong>color_map</strong> (<em>cmap object</em>) – color map to use for traces. Defaults to <cite>DEF_CMAP</cite>.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>figure object for plot with heatmap and dendrogram as described.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.plot.compartmentPlot">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.plot.</span></span><span class="sig-name descname"><span class="pre">compartmentPlot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hosts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vectors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_data_to_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Time'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Hosts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Compartment'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['#E69F00',</span> <span class="pre">'#56B4E9',</span> <span class="pre">'#009E73',</span> <span class="pre">'#F0E442',</span> <span class="pre">'#0072B2',</span> <span class="pre">'#D55E00',</span> <span class="pre">'#CC79A7',</span> <span class="pre">'#999999']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stacked</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/plot.html#compartmentPlot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.plot.compartmentPlot" title="Link to this definition"></a></dt>
<dd><p>Create plot with num. of naive, susc., inf., rec. hosts/vectors vs. time.</p>
<p>Creates a line or stacked line plot with dynamics of all compartments
(naive, infected, recovered, dead) across selected populations in the model,
with one line for each compartment.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> (<em>String</em>) – file path, name, and extension to save plot under.</p></li>
<li><p><strong>data</strong> (<em>pandas DataFrame</em>) – dataframe with model history as produced by <cite>saveToDf</cite> function.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>populations</strong> (<em>list</em><em> of </em><em>Strings</em>) – IDs of populations to include in analysis; if empty, uses all
populations in model. Defaults to [].</p></li>
<li><p><strong>hosts</strong> (<em>Boolean</em>) – whether to count hosts. Defaults to True.</p></li>
<li><p><strong>vectors</strong> (<em>Boolean</em>) – whether to count vectors. Defaults to False.</p></li>
<li><p><strong>save_data_to_file</strong> (<em>String</em>) – file path and name to save model data under, no saving
occurs if empty string. Defaults to “”.</p></li>
<li><p><strong>x_label</strong> (<em>String</em>) – X axis title. Defaults to ‘Time’.</p></li>
<li><p><strong>y_label</strong> (<em>String</em>) – Y axis title. Defaults to ‘Hosts’.</p></li>
<li><p><strong>legend_title</strong> (<em>String</em>) – legend title. Defaults to ‘Population’.</p></li>
<li><p><strong>legend_values</strong> (<em>list</em><em> of </em><em>Strings</em>) – labels for each trace, if empty list, uses population IDs.
Defaults to [].</p></li>
<li><p><strong>figsize</strong> (<em>array-like</em><em> of </em><em>two ints</em>) – dimensions of figure. Defaults to (8,4).</p></li>
<li><p><strong>dpi</strong> (<em>int</em>) – figure resolution. Defaults to 200.</p></li>
<li><p><strong>palette</strong> (<em>list</em><em> of </em><em>color Strings</em>) – color palette to use for traces. Defaults to <cite>CB_PALETTE</cite>.</p></li>
<li><p><strong>stacked</strong> (<em>Boolean</em>) – whether to draw a regular line plot instead of a stacked one.
Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>axis object for plot with model compartment dynamics as described above.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.plot.compositionPlot">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.plot.</span></span><span class="sig-name descname"><span class="pre">compositionPlot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">composition_dataframe</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type_of_composition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Pathogens'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hosts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vectors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_top_sequences</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">track_specific_sequences</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genomic_positions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count_individuals_based_on_model</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_data_to_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Time'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Infections'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Genotype'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['#E69F00',</span> <span class="pre">'#56B4E9',</span> <span class="pre">'#009E73',</span> <span class="pre">'#F0E442',</span> <span class="pre">'#0072B2',</span> <span class="pre">'#D55E00',</span> <span class="pre">'#CC79A7',</span> <span class="pre">'#999999']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stacked</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_legend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">population_fraction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/plot.html#compositionPlot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.plot.compositionPlot" title="Link to this definition"></a></dt>
<dd><p>Create plot with counts for pathogen genomes or resistance across time.</p>
<p>Creates a line or stacked line plot with dynamics of the pathogen strains or
protection sequences across selected populations in the model,
with one line for each pathogen genome or protection sequence being shown.</p>
<p>Of note: sum of totals for all sequences in one time point does not
necessarily equal the number of infected hosts and/or vectors, given
multiple infections in the same host/vector are counted separately.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> (<em>String</em>) – file path, name, and extension to save plot under.</p></li>
<li><p><strong>data</strong> (<em>pandas DataFrame</em>) – dataframe with model history as produced by <cite>saveToDf</cite> function.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>composition_dataframe</strong> (<em>Pandas DataFrame</em>) – output of compositionDf() if already computed.
Defaults to None.</p></li>
<li><p><strong>populations</strong> (<em>list</em><em> of </em><em>Strings</em>) – IDs of populations to include in analysis; if empty, uses all
populations in model. Defaults to [].</p></li>
<li><p><strong>type_of_composition</strong> (<em>String</em>) – field of data to count totals of, can be either
‘Pathogens’ or ‘Protection’. Defaults to ‘Pathogens’.</p></li>
<li><p><strong>hosts</strong> (<em>Boolean</em>) – whether to count hosts. Defaults to True.</p></li>
<li><p><strong>vectors</strong> (<em>Boolean</em>) – whether to count vectors. Defaults to False.</p></li>
<li><p><strong>num_top_sequences</strong> (<em>int</em>) – how many sequences to count separately and include
as columns, remainder will be counted under column “Other”; if &lt;0,
includes all genomes in model. Defaults to 7.</p></li>
<li><p><strong>track_specific_sequences</strong> (<em>list</em><em> of </em><em>Strings</em>) – contains specific sequences to have
as a separate column if not part of the top num_top_sequences
sequences. Defaults to [].</p></li>
<li><p><strong>genomic_positions</strong> (<em>list</em><em> of </em><em>lists</em><em> of </em><em>int</em>) – list in which each element is a list with loci
positions to extract (e.g. genomic_positions=[ [0,3], [5,6] ] extracts
positions 0, 1, 2, and 5 from each genome); if empty, takes full genomes. Defaults to [].</p></li>
<li><p><strong>count_individuals_based_on_model</strong> (<em>None</em><em> or </em><a class="reference internal" href="opqua.html#opqua.model.Model" title="opqua.model.Model"><em>Model</em></a>) – Model object with populations and
fitness functions used to evaluate the most fit pathogen genome in each
host/vector in order to count only a single pathogen per host/vector, as
opposed to all pathogens within each host/vector; if None, counts all
pathogens. Defaults to None.</p></li>
<li><p><strong>save_data_to_file</strong> (<em>String</em>) – file path and name to save model data under, no saving
occurs if empty string. Defaults to “”.</p></li>
<li><p><strong>x_label</strong> (<em>String</em>) – X axis title. Defaults to ‘Time’.</p></li>
<li><p><strong>y_label</strong> (<em>String</em>) – Y axis title. Defaults to ‘Hosts’.</p></li>
<li><p><strong>legend_title</strong> (<em>String</em>) – legend title. Defaults to ‘Population’.</p></li>
<li><p><strong>legend_values</strong> (<em>list</em><em> of </em><em>Strings</em>) – labels for each trace, if empty list, uses population IDs.
Defaults to [].</p></li>
<li><p><strong>figsize</strong> (<em>int</em>) – dimensions of figure. Defaults to (8,4).</p></li>
<li><p><strong>dpi</strong> (<em>int</em>) – figure resolution. Defaults to 200.</p></li>
<li><p><strong>palette</strong> (<em>list</em><em> of </em><em>color Strings</em>) – color palette to use for traces. Defaults to <cite>CB_PALETTE</cite>.</p></li>
<li><p><strong>stacked</strong> (<em>Boolean</em>) – whether to draw a regular line plot instead of a stacked one. Defaults to False.</p></li>
<li><p><strong>remove_legend</strong> (<em>Boolean</em>) – whether to print the sequences on the figure legend instead
of printing them on a separate csv file. Defaults to True.</p></li>
<li><p><strong>population_fraction</strong> (<em>Boolean</em>) – whether to graph fractions of pathogen population
instead of pathogen counts. Defaults to False.</p></li>
<li><p><strong>**kwargs</strong> – additional arguents for joblib multiprocessing.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>axis object for plot with model sequence composition dynamics as described.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="opqua.internal.plot.populationsPlot">
<span class="sig-prename descclassname"><span class="pre">opqua.internal.plot.</span></span><span class="sig-name descname"><span class="pre">populationsPlot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compartment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Infected'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hosts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vectors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_top_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">track_specific_populations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_data_to_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Time'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Infected</span> <span class="pre">hosts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Population'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">palette</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['#E69F00',</span> <span class="pre">'#56B4E9',</span> <span class="pre">'#009E73',</span> <span class="pre">'#F0E442',</span> <span class="pre">'#0072B2',</span> <span class="pre">'#D55E00',</span> <span class="pre">'#CC79A7',</span> <span class="pre">'#999999']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stacked</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/plot.html#populationsPlot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.plot.populationsPlot" title="Link to this definition"></a></dt>
<dd><p>Create plot with aggregated totals per population across time.</p>
<p>Creates a line or stacked line plot with dynamics of a compartment
across populations in the model, with one line for each population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> (<em>String</em>) – file path, name, and extension to save plot under.</p></li>
<li><p><strong>data</strong> (<em>pandas DataFrame</em>) – dataframe with model history as produced by <cite>saveToDf</cite> function.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>compartment</strong> (<em>String</em>) – subset of hosts/vectors to count totals of, can be either
‘Naive’,’Infected’,’Recovered’, or ‘Dead’. (default ‘Infected’)</p></li>
<li><p><strong>hosts</strong> (<em>Boolean</em>) – whether to count hosts. Defaults to True.</p></li>
<li><p><strong>vectors</strong> (<em>Boolean</em>) – whether to count vectors. Defaults to False.</p></li>
<li><p><strong>num_top_populations</strong> (<em>int</em>) – how many populations to count separately and include
as columns, remainder will be counted under column “Other”; if &lt;0,
includes all populations in model. Defaults to 7.</p></li>
<li><p><strong>track_specific_populations</strong> (<em>list</em><em> of </em><em>Strings</em>) – contains IDs of specific populations to have
as a separate column if not part of the top num_top_populations
populations. Defaults to [].</p></li>
<li><p><strong>save_data_to_file</strong> (<em>String</em>) – file path and name to save model plot data under, no
saving occurs if empty string. Defaults to “”.</p></li>
<li><p><strong>x_label</strong> (<em>String</em>) – X axis title. Defaults to ‘Time’.</p></li>
<li><p><strong>y_label</strong> (<em>String</em>) – Y axis title. Defaults to ‘Hosts’.</p></li>
<li><p><strong>legend_title</strong> (<em>String</em>) – legend title. Defaults to ‘Population’.</p></li>
<li><p><strong>legend_values</strong> (<em>list</em><em> of </em><em>Strings</em>) – labels for each trace, if empty list, uses population IDs.
Defaults to [].</p></li>
<li><p><strong>figsize</strong> (<em>array-like</em><em> of </em><em>two ints</em>) – dimensions of figure. Defaults to (8,4).</p></li>
<li><p><strong>dpi</strong> (<em>int</em>) – figure resolution. Defaults to 200.</p></li>
<li><p><strong>palette</strong> (<em>list</em><em> of </em><em>color Strings</em>) – color palette to use for traces. Defaults to <cite>CB_PALETTE</cite>.</p></li>
<li><p><strong>stacked</strong> (<em>Boolean</em>) – whether to draw a regular line plot instead of a stacked one. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>axis object for plot with model population dynamics as described above.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-opqua.internal.population">
<span id="opqua-internal-population-module"></span><h2>opqua.internal.population module<a class="headerlink" href="#module-opqua.internal.population" title="Link to this heading"></a></h2>
<p>Contains class Population.</p>
<dl class="py class">
<dt class="sig sig-object py" id="opqua.internal.population.Population">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">opqua.internal.population.</span></span><span class="sig-name descname"><span class="pre">Population</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">setup</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_hosts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_vectors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class defines a population with hosts, vectors, and specific parameters.</p>
<p><strong>CONSTANTS:</strong> These all denote positions in coefficients_hosts and coefficients_vectors</p>
<ul class="simple">
<li><dl class="simple">
<dt><cite>INFECTED</cite> – position of “infected” Boolean values for each individual inside</dt><dd><p>coefficients array.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>CONTACT</cite> – position of intra-population aggregated contact rate for each</dt><dd><p>individual inside coefficients array.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>RECEIVE_CONTACT</cite> – position of intra-population aggregated receiving contact</dt><dd><p>rate for each individual inside coefficients array.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>LETHALITY</cite> – position of aggregated death rate for each individual inside</dt><dd><p>coefficients array.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>NATALITY</cite> – position of aggregated birth rate for each individual inside</dt><dd><p>coefficients array.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>RECOVERY</cite> – position of aggregated recovery rate for each individual inside</dt><dd><p>coefficients array.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>MIGRATION</cite> – position of aggregated inter-population migration rate for each</dt><dd><p>individual inside coefficients array.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>POPULATION_CONTACT</cite> – position of inter-population aggregated contact rate</dt><dd><p>for each individual inside coefficients array.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>RECEIVE_POPULATION_CONTACT</cite> – position of inter-population aggregated</dt><dd><p>receiving contact rate for each individual inside coefficients array.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>MUTATION</cite> – position of aggregated mutation rate for each individual inside</dt><dd><p>coefficients array.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>RECOMBINATION</cite> – position of aggregated recovery rate for each individual</dt><dd><p>inside coefficients array.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>NUM_COEFFICIENTS</cite> – total number of types of coefficients (columns) in</dt><dd><p>coefficient arrays.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>CHROMOSOME_SEPARATOR</cite> – character reserved to denote separate chromosomes in</dt><dd><p>genomes.</p>
</dd>
</dl>
</li>
</ul>
<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.model">
<span class="sig-name descname"><span class="pre">model</span></span><a class="headerlink" href="#opqua.internal.population.Population.model" title="Link to this definition"></a></dt>
<dd><p>parent model this population is a part of.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Model object</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#opqua.internal.population.Population.id" title="Link to this definition"></a></dt>
<dd><p>unique identifier for this population in the model.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>String</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.setup">
<span class="sig-name descname"><span class="pre">setup</span></span><a class="headerlink" href="#opqua.internal.population.Population.setup" title="Link to this definition"></a></dt>
<dd><p>setup object with parameters for this population.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>String</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.num_hosts">
<span class="sig-name descname"><span class="pre">num_hosts</span></span><a class="headerlink" href="#opqua.internal.population.Population.num_hosts" title="Link to this definition"></a></dt>
<dd><p>number of hosts to initialize population with.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.num_vectors">
<span class="sig-name descname"><span class="pre">num_vectors</span></span><a class="headerlink" href="#opqua.internal.population.Population.num_vectors" title="Link to this definition"></a></dt>
<dd><p>number of hosts to initialize population with.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.slim">
<span class="sig-name descname"><span class="pre">slim</span></span><a class="headerlink" href="#opqua.internal.population.Population.slim" title="Link to this definition"></a></dt>
<dd><p>whether to create a slimmed-down representation of the
population for data storage (only ID, host and vector lists).
Defaults to False.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Boolean</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.CHROMOSOME_SEPARATOR">
<span class="sig-name descname"><span class="pre">CHROMOSOME_SEPARATOR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'/'</span></em><a class="headerlink" href="#opqua.internal.population.Population.CHROMOSOME_SEPARATOR" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.CONTACT">
<span class="sig-name descname"><span class="pre">CONTACT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1</span></em><a class="headerlink" href="#opqua.internal.population.Population.CONTACT" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.INFECTED">
<span class="sig-name descname"><span class="pre">INFECTED</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">0</span></em><a class="headerlink" href="#opqua.internal.population.Population.INFECTED" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.LETHALITY">
<span class="sig-name descname"><span class="pre">LETHALITY</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">3</span></em><a class="headerlink" href="#opqua.internal.population.Population.LETHALITY" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.MIGRATION">
<span class="sig-name descname"><span class="pre">MIGRATION</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">6</span></em><a class="headerlink" href="#opqua.internal.population.Population.MIGRATION" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.MUTATION">
<span class="sig-name descname"><span class="pre">MUTATION</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">9</span></em><a class="headerlink" href="#opqua.internal.population.Population.MUTATION" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.NATALITY">
<span class="sig-name descname"><span class="pre">NATALITY</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">4</span></em><a class="headerlink" href="#opqua.internal.population.Population.NATALITY" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.NUM_COEFFICIENTS">
<span class="sig-name descname"><span class="pre">NUM_COEFFICIENTS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">11</span></em><a class="headerlink" href="#opqua.internal.population.Population.NUM_COEFFICIENTS" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.POPULATION_CONTACT">
<span class="sig-name descname"><span class="pre">POPULATION_CONTACT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">7</span></em><a class="headerlink" href="#opqua.internal.population.Population.POPULATION_CONTACT" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.RECEIVE_CONTACT">
<span class="sig-name descname"><span class="pre">RECEIVE_CONTACT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">2</span></em><a class="headerlink" href="#opqua.internal.population.Population.RECEIVE_CONTACT" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.RECEIVE_POPULATION_CONTACT">
<span class="sig-name descname"><span class="pre">RECEIVE_POPULATION_CONTACT</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">8</span></em><a class="headerlink" href="#opqua.internal.population.Population.RECEIVE_POPULATION_CONTACT" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.RECOMBINATION">
<span class="sig-name descname"><span class="pre">RECOMBINATION</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">10</span></em><a class="headerlink" href="#opqua.internal.population.Population.RECOMBINATION" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.population.Population.RECOVERY">
<span class="sig-name descname"><span class="pre">RECOVERY</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">5</span></em><a class="headerlink" href="#opqua.internal.population.Population.RECOVERY" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.addHosts">
<span class="sig-name descname"><span class="pre">addHosts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_hosts</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.addHosts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.addHosts" title="Link to this definition"></a></dt>
<dd><p>Add a number of healthy hosts to population, return list with them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_hosts</strong> (<em>int</em>) – number of hosts to be added.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list containing new hosts.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.addPathogensToHosts">
<span class="sig-name descname"><span class="pre">addPathogensToHosts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genomes_numbers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hosts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.addPathogensToHosts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.addPathogensToHosts" title="Link to this definition"></a></dt>
<dd><p>Add specified pathogens to random hosts, optionally from a list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>genomes_numbers</strong> (<em>dict with keys=Strings</em><em>, </em><em>values=int</em>) – dictionary conatining
pathogen genomes to add as keys  and number of hosts each one will be
added to as values.</p>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>hosts</strong> (<em>list</em><em> of </em><em>Host objects</em>) – list of specific hosts to sample from, if empty, samples from
whole population. Defaults to [].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.addPathogensToVectors">
<span class="sig-name descname"><span class="pre">addPathogensToVectors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genomes_numbers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vectors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.addPathogensToVectors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.addPathogensToVectors" title="Link to this definition"></a></dt>
<dd><p>Add specified pathogens to random vectors, optionally from a list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>genomes_numbers</strong> (<em>dict with keys=Strings</em><em>, </em><em>values=int</em>) – dictionary conatining
pathogen genomes to add as keys and number of vectors each one will be
added to as values.</p>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>vectors</strong> (<em>list</em><em> of </em><em>Vector objects</em>) – list of specific vectors to sample from, if empty, samples
from whole population. Defaults to [].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.addVectors">
<span class="sig-name descname"><span class="pre">addVectors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_vectors</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.addVectors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.addVectors" title="Link to this definition"></a></dt>
<dd><p>Add a number of healthy vectors to population, return list with them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_vectors</strong> (<em>int</em>) – number of vectors to be added.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list containing new vectors</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.birthHost">
<span class="sig-name descname"><span class="pre">birthHost</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.birthHost"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.birthHost" title="Link to this definition"></a></dt>
<dd><p>Add host at this index to population, remove it from alive ones.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual to make parent.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.birthVector">
<span class="sig-name descname"><span class="pre">birthVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.birthVector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.birthVector" title="Link to this definition"></a></dt>
<dd><p>Add host at this index to population, remove it from alive ones.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual to make parent.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.contactHostHost">
<span class="sig-name descname"><span class="pre">contactHostHost</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.contactHostHost"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.contactHostHost" title="Link to this definition"></a></dt>
<dd><p>Contact any two (weighted) random hosts in population.</p>
<p>Carries out possible infection events from the first organism into the
second.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individuals to contact.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Boolean indicating whether or not the model has changed state.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.contactHostVector">
<span class="sig-name descname"><span class="pre">contactHostVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.contactHostVector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.contactHostVector" title="Link to this definition"></a></dt>
<dd><p>Contact a (weighted) random host and vector in population.</p>
<p>Carries out possible infection events from the first organism into the
second.</p>
<p>Arguments:
rand (number): uniform random number from 0 to 1 to use when choosing</p>
<blockquote>
<div><p>individuals to contact.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Boolean indicating whether or not the model has changed state.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.contactVectorHost">
<span class="sig-name descname"><span class="pre">contactVectorHost</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.contactVectorHost"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.contactVectorHost" title="Link to this definition"></a></dt>
<dd><p>Contact a (weighted) random vector and host in population.</p>
<p>Carries out possible infection events from the first organism into the
second.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – individuals to contact.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Boolean indicating whether or not the model has changed state.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.copyState">
<span class="sig-name descname"><span class="pre">copyState</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">host_sampling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vector_sampling</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.copyState"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.copyState" title="Link to this definition"></a></dt>
<dd><p>Returns a slimmed-down version of the current population state.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>host_sampling</strong> (<em>int</em>) – how many hosts to skip before saving one in a snapshot
of the system state (saves all by default). Defaults to 0.</p></li>
<li><p><strong>vector_sampling</strong> (<em>int</em>) – how many vectors to skip before saving one in a
snapshot of the system state (saves all by default). Defaults to 0.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Population object with current host and vector lists.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.dieHost">
<span class="sig-name descname"><span class="pre">dieHost</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.dieHost"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.dieHost" title="Link to this definition"></a></dt>
<dd><p>Remove host at this index from alive lists.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual to kill.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.dieVector">
<span class="sig-name descname"><span class="pre">dieVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.dieVector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.dieVector" title="Link to this definition"></a></dt>
<dd><p>Remove vector at this index from alive lists.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual to kill.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.getWeightedRandom">
<span class="sig-name descname"><span class="pre">getWeightedRandom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.getWeightedRandom"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.getWeightedRandom" title="Link to this definition"></a></dt>
<dd><p>Returns index of element chosen from weights and given random number.</p>
<p>Since sampling from coefficient arrays which contain a dummy first row,
index is decreased by 1.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rand</strong> (<em>number</em>) – 0-1 random number.</p></li>
<li><p><strong>r</strong> (<em>numpy array</em>) – array with weights.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>new 0-1 random number.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.healthyCoefficientRow">
<span class="sig-name descname"><span class="pre">healthyCoefficientRow</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.healthyCoefficientRow"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.healthyCoefficientRow" title="Link to this definition"></a></dt>
<dd><p>Returns coefficient values corresponding to a healthy host/vector.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.killHost">
<span class="sig-name descname"><span class="pre">killHost</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.killHost"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.killHost" title="Link to this definition"></a></dt>
<dd><p>Add host at this index to dead list, remove it from alive ones.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual to kill.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.killVector">
<span class="sig-name descname"><span class="pre">killVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.killVector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.killVector" title="Link to this definition"></a></dt>
<dd><p>Add host at this index to dead list, remove it from alive ones.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual to kill.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.migrate">
<span class="sig-name descname"><span class="pre">migrate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target_pop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_hosts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_vectors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.migrate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.migrate" title="Link to this definition"></a></dt>
<dd><p>Transfer hosts and/or vectors from this population to another.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>target_pop</strong> (<em>Population objects</em>) – population towards which migration will occur.</p></li>
<li><p><strong>num_hosts</strong> (<em>int</em>) – number of hosts to transfer.</p></li>
<li><p><strong>num_vectors</strong> (<em>int</em>) – number of vectors to transfer.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>rand</strong> (<em>number 0-1</em>) – uniform random number from 0 to 1 to use when choosing
individuals to migrate; if None, generates new random number to
choose (through numpy), otherwise, assumes event is happening
through Gillespie class call and migrates a single host or vector. Defaults to None.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.mutateHost">
<span class="sig-name descname"><span class="pre">mutateHost</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.mutateHost"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.mutateHost" title="Link to this definition"></a></dt>
<dd><p>Mutate a single, random locus in a random pathogen in the given host.</p>
<p>Creates a new genotype from a de novo mutation event in the host given.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual in which to choose a pathogen to mutate.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.mutateVector">
<span class="sig-name descname"><span class="pre">mutateVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.mutateVector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.mutateVector" title="Link to this definition"></a></dt>
<dd><p>Mutate a single, random locus in a random pathogen in given vector.</p>
<p>Creates a new genotype from a de novo mutation event in the vector
given.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual in which to choose a pathogen to mutate.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.newHostGroup">
<span class="sig-name descname"><span class="pre">newHostGroup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hosts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'any'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.newHostGroup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.newHostGroup" title="Link to this definition"></a></dt>
<dd><p>Return a list of random hosts in population.</p>
<dl class="field-list simple">
<dt class="field-odd">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hosts</strong> (<em>number</em>) – number of hosts to be sampled randomly: if &lt;0, samples from
whole population; if &lt;1, takes that fraction of population; if &gt;=1,
samples that integer number of hosts. Defaults to -1.</p></li>
<li><p><strong>type</strong> (<em>String = {'healthy'</em><em>, </em><em>'infected'</em><em>, </em><em>'any'}</em>) – whether to sample healthy hosts
only, infected hosts only, or any hosts. Defaults to ‘any’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list containing sampled hosts.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.newVectorGroup">
<span class="sig-name descname"><span class="pre">newVectorGroup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vectors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'any'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.newVectorGroup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.newVectorGroup" title="Link to this definition"></a></dt>
<dd><p>Return a list of random vectors in population.</p>
<dl class="field-list simple">
<dt class="field-odd">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>vectors</strong> (<em>number</em>) – number of vectors to be sampled randomly: if &lt;0, samples from
whole population; if &lt;1, takes that fraction of population; if &gt;=1,
samples that integer number of vectors. Defaults to -1.</p></li>
<li><p><strong>type</strong> (<em>String = {'healthy'</em><em>, </em><em>'infected'</em><em>, </em><em>'any'}</em>) – whether to sample healthy vectors
only, infected vectors. Defaults to ‘any’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>list containing sampled vectors.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.populationContact">
<span class="sig-name descname"><span class="pre">populationContact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target_pop</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">host_origin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">host_target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.populationContact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.populationContact" title="Link to this definition"></a></dt>
<dd><p>Contacts hosts and/or vectors from this population to another.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>target_pop</strong> (<em>Population object</em>) – population towards which migration will occur.</p></li>
<li><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individuals to contact.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>host_origin</strong> (<em>Boolean</em>) – whether to draw from hosts in the origin population
(as opposed to vectors). Defaults to True.</p></li>
<li><p><strong>host_target</strong> (<em>Boolean</em>) – whether to draw from hosts in the target population
(as opposed to vectors). Defaults to True.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.protectHosts">
<span class="sig-name descname"><span class="pre">protectHosts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frac_hosts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protection_sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hosts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.protectHosts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.protectHosts" title="Link to this definition"></a></dt>
<dd><p>Protect a random fraction of infected hosts against some infection.</p>
<p>Adds protection sequence specified to a random fraction of the hosts
specified. Does not cure them if they are already infected.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frac_hosts</strong> (<em>number 0-1</em>) – fraction of hosts considered to be randomly selected.</p></li>
<li><p><strong>protection_sequence</strong> (<em>String</em>) – sequence against which to protect.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>hosts</strong> (<em>list</em><em> of </em><em>Host objects</em>) – list of specific hosts to sample from, if empty, samples from
whole population. Defaults to [].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.protectVectors">
<span class="sig-name descname"><span class="pre">protectVectors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frac_vectors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protection_sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vectors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.protectVectors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.protectVectors" title="Link to this definition"></a></dt>
<dd><p>Protect a random fraction of infected vectors against some infection.</p>
<p>Adds protection sequence specified to a random fraction of the vectors
specified. Does not cure them if they are already infected.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frac_vectors</strong> (<em>number 0-1</em>) – fraction of vectors considered to be randomly selected.</p></li>
<li><p><strong>protection_sequence</strong> (<em>String</em>) – sequence against which to protect.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>vectors</strong> (<em>list</em><em> of </em><em>Vector objects</em>) – list of specific vectors to sample from, if empty, samples
from whole population. Defaults to [].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.recombineHost">
<span class="sig-name descname"><span class="pre">recombineHost</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.recombineHost"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.recombineHost" title="Link to this definition"></a></dt>
<dd><p>Recombine 2 random pathogen genomes at random locus in given host.</p>
<p>Creates a new genotype from two random possible pathogens in the host
given.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual in which to choose pathogens to recombine.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.recombineVector">
<span class="sig-name descname"><span class="pre">recombineVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.recombineVector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.recombineVector" title="Link to this definition"></a></dt>
<dd><p>Recombine 2 random pathogen genomes at random locus in given vector.</p>
<p>Creates a new genotype from two random possible pathogens in the vector
given.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual in which to choose pathogens to recombine.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.recoverHost">
<span class="sig-name descname"><span class="pre">recoverHost</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.recoverHost"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.recoverHost" title="Link to this definition"></a></dt>
<dd><p>Remove all infections from host at this index.</p>
<p>If model is protecting upon recovery, add protecion sequence as defined
by the indexes in the corresponding model parameter. Remove from
population infected list and add to healthy list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual to recover.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.recoverVector">
<span class="sig-name descname"><span class="pre">recoverVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.recoverVector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.recoverVector" title="Link to this definition"></a></dt>
<dd><p>Remove all infections from vector at this index.</p>
<p>If model is protecting upon recovery, add protecion sequence as defined
by the indexes in the corresponding model parameter. Remove from
population infected list and add to healthy list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>rand</strong> (<em>number</em>) – uniform random number from 0 to 1 to use when choosing
individual to recover.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.removeHosts">
<span class="sig-name descname"><span class="pre">removeHosts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_hosts_or_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.removeHosts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.removeHosts" title="Link to this definition"></a></dt>
<dd><p>Remove a number of specified or random hosts from population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_hosts_or_list</strong> (<em>int</em><em> or </em><em>list</em><em> of </em><em>Host objects</em>) – number of hosts to be sampled randomly for removal
or list of hosts to be removed, must be hosts in this population.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.removeVectors">
<span class="sig-name descname"><span class="pre">removeVectors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_vectors_or_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.removeVectors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.removeVectors" title="Link to this definition"></a></dt>
<dd><p>Remove a number of specified or random vectors from population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>num_vectors_or_list</strong> (<em>int</em><em> or </em><em>list</em><em> of </em><em>Vector objects</em>) – number of vectors to be sampled randomly for
removal or list of vectors to be removed, must be vectors in this
population.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.setHostHostPopulationContactNeighbor">
<span class="sig-name descname"><span class="pre">setHostHostPopulationContactNeighbor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">neighbor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.setHostHostPopulationContactNeighbor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.setHostHostPopulationContactNeighbor" title="Link to this definition"></a></dt>
<dd><p>Set host-host contact rate from this population towards another one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>neighbor</strong> (<em>Population object</em>) – population towards which migration rate will be specified.</p></li>
<li><p><strong>rate</strong> (<em>number</em>) – migration rate from this population to the neighbor.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.setHostMigrationNeighbor">
<span class="sig-name descname"><span class="pre">setHostMigrationNeighbor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">neighbor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.setHostMigrationNeighbor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.setHostMigrationNeighbor" title="Link to this definition"></a></dt>
<dd><p>Set host migration rate from this population towards another one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>neighbor</strong> (<em>Population object</em>) – population towards which migration rate will be specified.</p></li>
<li><p><strong>rate</strong> (<em>number</em>) – migration rate from this population to the neighbor.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.setHostVectorPopulationContactNeighbor">
<span class="sig-name descname"><span class="pre">setHostVectorPopulationContactNeighbor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">neighbor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.setHostVectorPopulationContactNeighbor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.setHostVectorPopulationContactNeighbor" title="Link to this definition"></a></dt>
<dd><p>Set host-vector contact rate from this population to another one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>neighbor</strong> (<em>Population object</em>) – population towards which migration rate will be specified.</p></li>
<li><p><strong>rate</strong> (<em>number</em>) – migration rate from this population to the neighbor.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.setSetup">
<span class="sig-name descname"><span class="pre">setSetup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">setup</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.setSetup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.setSetup" title="Link to this definition"></a></dt>
<dd><p>Assign parameters stored in Setup object to this population.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>setup</strong> (<em>Setup object</em>) – the setup to be assigned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.setVectorHostPopulationContactNeighbor">
<span class="sig-name descname"><span class="pre">setVectorHostPopulationContactNeighbor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">neighbor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.setVectorHostPopulationContactNeighbor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.setVectorHostPopulationContactNeighbor" title="Link to this definition"></a></dt>
<dd><p>Set vector-host contact rate from this population to another one.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>neighbor</strong> (<em>Population object</em>) – population towards which migration rate will be specified.</p></li>
<li><p><strong>rate</strong> (<em>number</em>) – migration rate from this population to the neighbor.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.setVectorMigrationNeighbor">
<span class="sig-name descname"><span class="pre">setVectorMigrationNeighbor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">neighbor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rate</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.setVectorMigrationNeighbor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.setVectorMigrationNeighbor" title="Link to this definition"></a></dt>
<dd><p>Set vector migration rate from this population towards another one.</p>
<p>Arguments:
neighbor (Population object): population towards which migration rate will be specified.
rate (number): migration rate from this population to the neighbor.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.treatHosts">
<span class="sig-name descname"><span class="pre">treatHosts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frac_hosts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resistance_seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hosts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.treatHosts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.treatHosts" title="Link to this definition"></a></dt>
<dd><p>Treat random fraction of infected hosts against some infection.</p>
<p>Removes all infections with genotypes susceptible to given treatment.
Pathogens are removed if they are missing at least one of the sequences
in resistance_seqs from their genome. Removes this organism from
population infected list and adds to healthy list if appropriate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frac_hosts</strong> (<em>number 0-1</em>) – fraction of hosts considered to be randomly selected.</p></li>
<li><p><strong>resistance_seqs</strong> (<em>list</em><em> of </em><em>Strings</em>) – contains sequences required for treatment resistance.</p></li>
</ul>
</dd>
</dl>
<p>Keyword arguments:
hosts (list of Host objects): list of specific hosts to sample from, if empty, samples from</p>
<blockquote>
<div><p>whole population. Defaults to [].</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.treatVectors">
<span class="sig-name descname"><span class="pre">treatVectors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">frac_vectors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resistance_seqs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vectors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.treatVectors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.treatVectors" title="Link to this definition"></a></dt>
<dd><p>Treat random fraction of infected vectors agains some infection.</p>
<p>Removes all infections with genotypes susceptible to given treatment.
Pathogens are removed if they are missing at least one of the sequences
in resistance_seqs from their genome. Removes this organism from
population infected list and adds to healthy list if appropriate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>frac_vectors</strong> (<em>number 0-1</em>) – fraction of vectors considered to be randomly selected.</p></li>
<li><p><strong>resistance_seqs</strong> (<em>list</em><em> of </em><em>Strings</em>) – contains sequences required for treatment resistance.</p></li>
</ul>
</dd>
<dt class="field-even">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>vectors</strong> (<em>list</em><em> of </em><em>Vector objects</em>) – list of specific vectors to sample from, if empty, samples
from whole population. Defaults to [].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.updateHostCoefficients">
<span class="sig-name descname"><span class="pre">updateHostCoefficients</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.updateHostCoefficients"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.updateHostCoefficients" title="Link to this definition"></a></dt>
<dd><p>Updates event coefficient values in population’s hosts.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.updateVectorCoefficients">
<span class="sig-name descname"><span class="pre">updateVectorCoefficients</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.updateVectorCoefficients"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.updateVectorCoefficients" title="Link to this definition"></a></dt>
<dd><p>Updates event coefficient values in population’s vectors.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.wipeProtectionHosts">
<span class="sig-name descname"><span class="pre">wipeProtectionHosts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hosts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.wipeProtectionHosts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.wipeProtectionHosts" title="Link to this definition"></a></dt>
<dd><p>Removes all protection sequences from hosts.</p>
<dl class="field-list simple">
<dt class="field-odd">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>hosts</strong> (<em>list</em><em> of </em><em>Host objects</em>) – list of specific hosts to sample from, if empty, samples from
whole population. Defaults to [].</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.population.Population.wipeProtectionVectors">
<span class="sig-name descname"><span class="pre">wipeProtectionVectors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vectors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/population.html#Population.wipeProtectionVectors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.population.Population.wipeProtectionVectors" title="Link to this definition"></a></dt>
<dd><p>Removes all protection sequences from vectors.</p>
<dl class="field-list simple">
<dt class="field-odd">Keyword Arguments<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vectors</strong> (<em>list</em><em> of </em><em>Vector objects</em>) – list of specific vectors to sample from, if empty, samples from
whole population. Defaults to [].</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-opqua.internal.setup">
<span id="opqua-internal-setup-module"></span><h2>opqua.internal.setup module<a class="headerlink" href="#module-opqua.internal.setup" title="Link to this heading"></a></h2>
<p>Contains class Intervention.</p>
<dl class="py class">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">opqua.internal.setup.</span></span><span class="sig-name descname"><span class="pre">Setup</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_loci</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">possible_alleles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitnessHost</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contactHost</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receiveContactHost</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortalityHost</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">natalityHost</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recoveryHost</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">migrationHost</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populationContactHost</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receivePopulationContactHost</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutationHost</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombinationHost</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fitnessVector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contactVector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receiveContactVector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortalityVector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">natalityVector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recoveryVector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">migrationVector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">populationContactVector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receivePopulationContactVector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutationVector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombinationVector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contact_rate_host_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transmission_efficiency_host_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transmission_efficiency_vector_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contact_rate_host_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transmission_efficiency_host_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_inoculum_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_inoculum_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recovery_rate_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recovery_rate_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortality_rate_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mortality_rate_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombine_in_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">recombine_in_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_crossover_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_crossover_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutate_in_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutate_in_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">death_rate_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">death_rate_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">birth_rate_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">birth_rate_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertical_transmission_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertical_transmission_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inherit_protection_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inherit_protection_vector</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protection_upon_recovery_host</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protection_upon_recovery_vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/setup.html#Setup"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.setup.Setup" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class defines a setup with population parameters.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.id" title="Link to this definition"></a></dt>
<dd><p>key of the Setup inside model dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>String</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.num_loci">
<span class="sig-name descname"><span class="pre">num_loci</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.num_loci" title="Link to this definition"></a></dt>
<dd><p>length of each pathogen genome string.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int&gt;0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.possible_alleles">
<span class="sig-name descname"><span class="pre">possible_alleles</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.possible_alleles" title="Link to this definition"></a></dt>
<dd><p>set of possible
characters in all genome string, or at each position in genome string.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>String or list of Strings with num_loci elements</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.fitnessHost">
<span class="sig-name descname"><span class="pre">fitnessHost</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.fitnessHost" title="Link to this definition"></a></dt>
<dd><p>function
that evaluates relative fitness in head-to-head competition for different genomes
within the same host.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.contactHost">
<span class="sig-name descname"><span class="pre">contactHost</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.contactHost" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying probability of a given host being chosen to be the
infector in a contact event, based on genome sequence of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.receiveContactHost">
<span class="sig-name descname"><span class="pre">receiveContactHost</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.receiveContactHost" title="Link to this definition"></a></dt>
<dd><p>function
that returns coefficient modifying probability of a given host being chosen to be
the infected in a contact event, based on genome sequence of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.mortalityHost">
<span class="sig-name descname"><span class="pre">mortalityHost</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.mortalityHost" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying death rate for a given host, based on genome sequence
of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.natalityHost">
<span class="sig-name descname"><span class="pre">natalityHost</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.natalityHost" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying birth rate for a given host, based on genome sequence
of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.recoveryHost">
<span class="sig-name descname"><span class="pre">recoveryHost</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.recoveryHost" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying recovery rate for a given host based on genome sequence
of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.migrationHost">
<span class="sig-name descname"><span class="pre">migrationHost</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.migrationHost" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying migration rate for a given host based on genome sequence
of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.populationContactHost">
<span class="sig-name descname"><span class="pre">populationContactHost</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.populationContactHost" title="Link to this definition"></a></dt>
<dd><p>function
that returns coefficient modifying population contact rate for a given host based on
genome sequence of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.mutationHost">
<span class="sig-name descname"><span class="pre">mutationHost</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.mutationHost" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying mutation rate for a given host based on genome sequence
of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.recombinationHost">
<span class="sig-name descname"><span class="pre">recombinationHost</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.recombinationHost" title="Link to this definition"></a></dt>
<dd><p>function
that returns coefficient modifying recombination rate for a given host based on genome
sequence of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.fitnessVector">
<span class="sig-name descname"><span class="pre">fitnessVector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.fitnessVector" title="Link to this definition"></a></dt>
<dd><p>function that
evaluates relative fitness in head-to-head competition for different genomes within
the same vector.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number &gt;=0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.contactVector">
<span class="sig-name descname"><span class="pre">contactVector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.contactVector" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying probability of a given vector being chosen to be the
infector in a contact event, based on genome sequence of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.receiveContactVector">
<span class="sig-name descname"><span class="pre">receiveContactVector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.receiveContactVector" title="Link to this definition"></a></dt>
<dd><p>function
that returns coefficient modifying probability of a given vector being chosen to be the
infected in a contact event, based on genome sequence of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.mortalityVector">
<span class="sig-name descname"><span class="pre">mortalityVector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.mortalityVector" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying death rate for a given vector, based on genome sequence
of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.natalityVector">
<span class="sig-name descname"><span class="pre">natalityVector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.natalityVector" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying birth rate for a given vector, based on genome sequence
of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.recoveryVector">
<span class="sig-name descname"><span class="pre">recoveryVector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.recoveryVector" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying recovery rate for a given vector based on genome sequence
of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.migrationVector">
<span class="sig-name descname"><span class="pre">migrationVector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.migrationVector" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying migration rate for a given vector based on genome sequence
of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.populationContactVector">
<span class="sig-name descname"><span class="pre">populationContactVector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.populationContactVector" title="Link to this definition"></a></dt>
<dd><p>function
that returns coefficient modifying population contact rate for a given vector based on
genome sequence of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.mutationVector">
<span class="sig-name descname"><span class="pre">mutationVector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.mutationVector" title="Link to this definition"></a></dt>
<dd><p>function that
returns coefficient modifying mutation rate for a given vector based on genome sequence
of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.recombinationVector">
<span class="sig-name descname"><span class="pre">recombinationVector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.recombinationVector" title="Link to this definition"></a></dt>
<dd><p>function
that returns coefficient modifying recombination rate for a given vector based on genome
sequence of pathogen.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>callable, takes a String argument and returns a number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.contact_rate_host_vector">
<span class="sig-name descname"><span class="pre">contact_rate_host_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.contact_rate_host_vector" title="Link to this definition"></a></dt>
<dd><p>rate of host-vector contact events, not necessarily
transmission, assumes constant population density; evts/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.transmission_efficiency_host_vector">
<span class="sig-name descname"><span class="pre">transmission_efficiency_host_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.transmission_efficiency_host_vector" title="Link to this definition"></a></dt>
<dd><p>fraction of host-vector contacts
that result in successful transmission.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.transmission_efficiency_vector_host">
<span class="sig-name descname"><span class="pre">transmission_efficiency_vector_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.transmission_efficiency_vector_host" title="Link to this definition"></a></dt>
<dd><p>fraction of vector-host contacts
that result in successful transmission.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.contact_rate_host_host">
<span class="sig-name descname"><span class="pre">contact_rate_host_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.contact_rate_host_host" title="Link to this definition"></a></dt>
<dd><p>rate of host-host contact events, not
necessarily transmission, assumes constant population density; evts/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.transmission_efficiency_host_host">
<span class="sig-name descname"><span class="pre">transmission_efficiency_host_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.transmission_efficiency_host_host" title="Link to this definition"></a></dt>
<dd><p>fraction of host-host contacts
that result in successful transmission.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.mean_inoculum_host">
<span class="sig-name descname"><span class="pre">mean_inoculum_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.mean_inoculum_host" title="Link to this definition"></a></dt>
<dd><p>mean number of pathogens that are transmitted from
a vector or host into a new host during a contact event.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.mean_inoculum_vector">
<span class="sig-name descname"><span class="pre">mean_inoculum_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.mean_inoculum_vector" title="Link to this definition"></a></dt>
<dd><p>from a host to a vector during a contact event.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.recovery_rate_host">
<span class="sig-name descname"><span class="pre">recovery_rate_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.recovery_rate_host" title="Link to this definition"></a></dt>
<dd><p>rate at which hosts clear all pathogens;
1/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.recovery_rate_vector">
<span class="sig-name descname"><span class="pre">recovery_rate_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.recovery_rate_vector" title="Link to this definition"></a></dt>
<dd><p>rate at which vectors clear all pathogens
1/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id0">
<span class="sig-name descname"><span class="pre">recovery_rate_vector</span></span><a class="headerlink" href="#id0" title="Link to this definition"></a></dt>
<dd><p>rate at which vectors clear all pathogens
1/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.mortality_rate_host">
<span class="sig-name descname"><span class="pre">mortality_rate_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.mortality_rate_host" title="Link to this definition"></a></dt>
<dd><p>rate at which infected hosts die from disease.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.mortality_rate_vector">
<span class="sig-name descname"><span class="pre">mortality_rate_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.mortality_rate_vector" title="Link to this definition"></a></dt>
<dd><p>rate at which infected vectors die from
disease.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.recombine_in_host">
<span class="sig-name descname"><span class="pre">recombine_in_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.recombine_in_host" title="Link to this definition"></a></dt>
<dd><p>rate at which recombination occurs in host;
evts/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.recombine_in_vector">
<span class="sig-name descname"><span class="pre">recombine_in_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.recombine_in_vector" title="Link to this definition"></a></dt>
<dd><p>rate at which recombination occurs in vector;
evts/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.num_crossover_host">
<span class="sig-name descname"><span class="pre">num_crossover_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.num_crossover_host" title="Link to this definition"></a></dt>
<dd><p>mean of a Poisson distribution modeling the number
of crossover events of host recombination events.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.num_crossover_vector">
<span class="sig-name descname"><span class="pre">num_crossover_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.num_crossover_vector" title="Link to this definition"></a></dt>
<dd><p>mean of a Poisson distribution modeling the
number of crossover events of vector recombination events.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.mutate_in_host">
<span class="sig-name descname"><span class="pre">mutate_in_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.mutate_in_host" title="Link to this definition"></a></dt>
<dd><p>rate at which mutation occurs in host; evts/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.mutate_in_vector">
<span class="sig-name descname"><span class="pre">mutate_in_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.mutate_in_vector" title="Link to this definition"></a></dt>
<dd><p>rate at which mutation occurs in vector; evts/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.death_rate_host">
<span class="sig-name descname"><span class="pre">death_rate_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.death_rate_host" title="Link to this definition"></a></dt>
<dd><p>natural host death rate; 1/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.death_rate_vector">
<span class="sig-name descname"><span class="pre">death_rate_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.death_rate_vector" title="Link to this definition"></a></dt>
<dd><p>natural vector death rate; 1/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.birth_rate_host">
<span class="sig-name descname"><span class="pre">birth_rate_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.birth_rate_host" title="Link to this definition"></a></dt>
<dd><p>infected host birth rate; 1/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.birth_rate_vector">
<span class="sig-name descname"><span class="pre">birth_rate_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.birth_rate_vector" title="Link to this definition"></a></dt>
<dd><p>infected vector birth rate; 1/time.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number &gt;= 0</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.vertical_transmission_host">
<span class="sig-name descname"><span class="pre">vertical_transmission_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.vertical_transmission_host" title="Link to this definition"></a></dt>
<dd><p>probability that a host is infected by its
parent at birth.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.vertical_transmission_vector">
<span class="sig-name descname"><span class="pre">vertical_transmission_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.vertical_transmission_vector" title="Link to this definition"></a></dt>
<dd><p>probability that a vector is infected by
its parent at birth.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.inherit_protection_host">
<span class="sig-name descname"><span class="pre">inherit_protection_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.inherit_protection_host" title="Link to this definition"></a></dt>
<dd><p>probability that a host inherits all
protection sequences from its parent.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.inherit_protection_vector">
<span class="sig-name descname"><span class="pre">inherit_protection_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.inherit_protection_vector" title="Link to this definition"></a></dt>
<dd><p>probability that a vector inherits all
protection sequences from its parent.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>number 0-1</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.protection_upon_recovery_host">
<span class="sig-name descname"><span class="pre">protection_upon_recovery_host</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.protection_upon_recovery_host" title="Link to this definition"></a></dt>
<dd><p>defines
indexes in genome string that define substring to be added to host protection sequences
after recovery.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None or array-like of length 2 with int 0-num_loci</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.setup.Setup.protection_upon_recovery_vector">
<span class="sig-name descname"><span class="pre">protection_upon_recovery_vector</span></span><a class="headerlink" href="#opqua.internal.setup.Setup.protection_upon_recovery_vector" title="Link to this definition"></a></dt>
<dd><p>defines
indexes in genome string that define substring to be added to vector protection sequences
after recovery.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None or array-like of length 2 with int 0-num_loci</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-opqua.internal.vector">
<span id="opqua-internal-vector-module"></span><h2>opqua.internal.vector module<a class="headerlink" href="#module-opqua.internal.vector" title="Link to this heading"></a></h2>
<p>Contains class Vector.</p>
<dl class="py class">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">opqua.internal.vector.</span></span><span class="sig-name descname"><span class="pre">Vector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">population</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Class defines vector entities to be infected by pathogens in model.</p>
<p>These can infect hosts, the main entities in the model.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.population">
<span class="sig-name descname"><span class="pre">population</span></span><a class="headerlink" href="#opqua.internal.vector.Vector.population" title="Link to this definition"></a></dt>
<dd><p>the population this vector belongs to.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Population object</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.id">
<span class="sig-name descname"><span class="pre">id</span></span><a class="headerlink" href="#opqua.internal.vector.Vector.id" title="Link to this definition"></a></dt>
<dd><p>unique identifier for this vector within population.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>String</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.slim">
<span class="sig-name descname"><span class="pre">slim</span></span><a class="headerlink" href="#opqua.internal.vector.Vector.slim" title="Link to this definition"></a></dt>
<dd><p>whether to create a slimmed-down representation of the
population for data storage (only ID, host and vector lists). Defaults to False.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Boolean</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.acquirePathogen">
<span class="sig-name descname"><span class="pre">acquirePathogen</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector.acquirePathogen"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector.acquirePathogen" title="Link to this definition"></a></dt>
<dd><p>Adds given genome to this vector’s pathogens.</p>
<p>Modifies event coefficient matrix accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>genome</strong> (<em>String</em>) – the genome to be added.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Boolean indicating whether or not the model has changed state.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.applyTreatment">
<span class="sig-name descname"><span class="pre">applyTreatment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">resistance_seqs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector.applyTreatment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector.applyTreatment" title="Link to this definition"></a></dt>
<dd><p>Remove all infections with genotypes susceptible to given treatment.</p>
<p>Pathogens are removed if they are missing at least one of the sequences
in resistance_seqs from their genome. Removes this organism from
population infected list and adds to healthy list if appropriate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>resistance_seqs</strong> (<em>list</em><em> of </em><em>Strings</em>) – contains sequences required for treatment resistance.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.birth">
<span class="sig-name descname"><span class="pre">birth</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector.birth"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector.birth" title="Link to this definition"></a></dt>
<dd><p>Add vector to population based on this vector.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.copyState">
<span class="sig-name descname"><span class="pre">copyState</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector.copyState"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector.copyState" title="Link to this definition"></a></dt>
<dd><p>Returns a slimmed-down representation of the current vector state.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Vector object with current pathogens and protection_sequences.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.die">
<span class="sig-name descname"><span class="pre">die</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector.die"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector.die" title="Link to this definition"></a></dt>
<dd><p>Add vector to population’s dead list, remove it from alive ones.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.getWeightedRandomGenome">
<span class="sig-name descname"><span class="pre">getWeightedRandomGenome</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">r</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector.getWeightedRandomGenome"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector.getWeightedRandomGenome" title="Link to this definition"></a></dt>
<dd><p>Returns index of element chosen from weights and given random number.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rand</strong> (<em>number</em>) – 0-1 random number.</p></li>
<li><p><strong>r</strong> (<em>numpy array</em>) – array with weights.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>new 0-1 random number.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.infectHost">
<span class="sig-name descname"><span class="pre">infectHost</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">host</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector.infectHost"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector.infectHost" title="Link to this definition"></a></dt>
<dd><p>Infect given host with a sample of this vector’s pathogens.</p>
<p>Each pathogen in the infector is sampled as present or absent in the
inoculum by drawing from a Poisson distribution with a mean equal to the
mean inoculum size of the organism being infected weighted by each
genome’s fitness as a fraction of the total in the infector as the
probability of each trial (minimum 1 pathogen transfered). Each pathogen
present in the inoculum will be added to the infected organism, if it
does not have protection from the pathogen’s genome. Fitnesses are
computed for the pathogens’ genomes in the infected organism, and the
organism is included in the poplation’s infected list if appropriate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vector</strong> (<em>Vector object</em>) – the vector to be infected.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Boolean indicating whether or not the model has changed state.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.infectVector">
<span class="sig-name descname"><span class="pre">infectVector</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector.infectVector"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector.infectVector" title="Link to this definition"></a></dt>
<dd><p>Infect given host with a sample of this vector’s pathogens.</p>
<p>Each pathogen in the infector is sampled as present or absent in the
inoculum by drawing from a Poisson distribution with a mean equal to the
mean inoculum size of the organism being infected weighted by each
genome’s fitness as a fraction of the total in the infector as the
probability of each trial (minimum 1 pathogen transfered). Each pathogen
present in the inoculum will be added to the infected organism, if it
does not have protection from the pathogen’s genome. Fitnesses are
computed for the pathogens’ genomes in the infected organism, and the
organism is included in the poplation’s infected list if appropriate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vector</strong> (<em>Vector object</em>) – the vector to be infected.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Boolean indicating whether or not the model has changed state.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.mutate">
<span class="sig-name descname"><span class="pre">mutate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector.mutate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector.mutate" title="Link to this definition"></a></dt>
<dd><p>Mutate a single, random locus in a random pathogen.</p>
<p>Creates a new genotype from a de novo mutation event.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.recombine">
<span class="sig-name descname"><span class="pre">recombine</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rand</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector.recombine"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector.recombine" title="Link to this definition"></a></dt>
<dd><p>Recombine two random pathogen genomes at random locus.</p>
<p>Creates a new genotype from two random possible pathogens.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="opqua.internal.vector.Vector.recover">
<span class="sig-name descname"><span class="pre">recover</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/opqua/internal/vector.html#Vector.recover"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#opqua.internal.vector.Vector.recover" title="Link to this definition"></a></dt>
<dd><p>Remove all infections from this vector.</p>
<p>If model is protecting upon recovery, add protection sequence as defined
by the indexes in the corresponding model parameter. Remove from
population infected list and add to healthy list.</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-opqua.internal">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-opqua.internal" title="Link to this heading"></a></h2>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="opqua.html" class="btn btn-neutral float-left" title="opqua package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Pablo Cárdenas.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>